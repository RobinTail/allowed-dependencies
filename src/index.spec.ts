import { afterAll, describe, expect, it } from "bun:test";
import parser from "@typescript-eslint/parser";
import { RuleTester } from "@typescript-eslint/rule-tester";
import plugin from "./index";

describe("Allowed dependencies", () => {
  RuleTester.afterAll = afterAll;
  RuleTester.describe = describe;
  RuleTester.it = it;
  const tester = new RuleTester({
    languageOptions: { parser },
  });

  it("should consist of one rule", () => {
    expect(plugin).toMatchSnapshot();
  });

  tester.run("dependencies", plugin.rules.dependencies, {
    valid: [
      {
        name: "regular import of prod dependency",
        code: `import {} from "eslint"`,
        options: [{ manifest: { dependencies: { eslint: "" } } }],
      },
      {
        name: "regular import of required peer dependency",
        code: `import {} from "eslint"`,
        options: [{ manifest: { peerDependencies: { eslint: "" } } }],
      },
      {
        name: "type import of optional peer dependency",
        code: `import type {} from "eslint"`,
        options: [
          {
            manifest: {
              peerDependencies: { eslint: "" },
              peerDependenciesMeta: { eslint: { optional: true } },
            },
          },
        ],
      },
      {
        name: "type import of unlisted type only dependency",
        code: `import type {} from "eslint"`,
        options: [{ manifest: { dependencies: {} }, typeOnly: ["eslint"] }],
      },
    ],
    invalid: [
      {
        name: "regular import of unlisted dependency",
        code: `import {} from "eslint"`,
        options: [{ manifest: { dependencies: {} } }],
        errors: [{ messageId: "typeOnly" }], // @todo should be prohibited
      },
    ],
  });
});
