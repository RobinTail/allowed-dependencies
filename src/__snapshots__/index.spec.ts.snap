// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`Allowed dependencies should consist of one rule 1`] = `
{
  "rules": {
    "dependencies": {
      "create": [Function: create],
      "defaultOptions": [
        {
          "manifest": {},
          "optionalPeers": "typeOnly",
          "production": true,
          "requiredPeers": true,
        },
      ],
      "meta": {
        "messages": {
          "prohibited": "Importing {{name}} is not allowed.",
          "typeOnly": "Only 'import type' syntax is allowed for {{name}}.",
        },
        "schema": [
          {
            "additionalProperties": false,
            "properties": {
              "manifest": {
                "properties": {
                  "dependencies": {
                    "type": "object",
                  },
                  "peerDependencies": {
                    "type": "object",
                  },
                  "peerDependenciesMeta": {
                    "type": "object",
                  },
                },
                "type": "object",
              },
              "optionalPeers": {
                "oneOf": [
                  {
                    "type": "boolean",
                  },
                  {
                    "enum": [
                      "typeOnly",
                    ],
                    "type": "string",
                  },
                ],
              },
              "production": {
                "oneOf": [
                  {
                    "type": "boolean",
                  },
                  {
                    "enum": [
                      "typeOnly",
                    ],
                    "type": "string",
                  },
                ],
              },
              "requiredPeers": {
                "oneOf": [
                  {
                    "type": "boolean",
                  },
                  {
                    "enum": [
                      "typeOnly",
                    ],
                    "type": "string",
                  },
                ],
              },
              "typeOnly": {
                "items": {
                  "type": "string",
                },
                "type": "array",
              },
            },
            "required": [
              "manifest",
            ],
            "type": "object",
          },
        ],
        "type": "problem",
      },
    },
  },
}
`;
